README.rst
``blanco`` - *"Hey, remember me?"*
==================================

Introduction
------------

``blanco`` is a simple tool to help you, or more specifically *me*, keep in
touch with people.  All it does is notify you if you’re failing to keep in
contact.  It is just a quick solution to a simple problem, as long as you use
abook_ and your sent mail is locally accessible.

It is written in Python_ and licensed under the `GPL v3`_.

Configuration
-------------

``blanco`` expects your abook_ entries to have a frequency value in the
``frequency`` field [#]_.  The format is “<n> <units>”, where ``n`` is a number
and ``units`` is a character from the set ``[dwmy]``.  For example, an entry
with a frequency of ``3 m`` will be triggered if there hasn’t been a mail sent
to that address in three months.

If you add the following snippet to your ``~/.abook/abookrc`` the abook_
interface will display “Frequency” in the interface when editing the ``other``
tab for a contact.

::

    field frequency = Frequency

    view OTHER = frequency

.. [#] You can select a different field using the ``--field`` option to
       ``blanco``.

Example
-------

``blanco`` expects you to give it the location of an abook_ addressbook file and
the location of your sent mail in mbox format, and informs you if you’re
forgetting to contact somebody.

::

    $ ./blanco.py --no-notify -a tests/data/blanco.conf -m tests/data/sent.mbox
    Mail due for Bill
    Mail due for Joe
    No mail record for Steven
    $ ./blanco.py --no-notify -a tests/data/blanco.conf -m tests/data/sent.mbox --all
    Mail due for Bill
    Mail due for Joe
    Mail due for Steven

Without abook
'''''''''''''

``blanco`` can be used without abook_, as it only requires a ini_ formatted
contacts file and access to a mailbox containing sent mail.  To create your own
contacts file without abook follow the format below::

    [0]
    name=Bill
    email=test@example.com
    frequency=30d

    [1]
    name=Joe
    email=joe@example.com
    frequency=30d

If you use the layout above you should specify ``--field=frequency`` when
calling ``blanco``.

Another alternative would be to use abook_ just to convert your current address
book in to a suitable format.  Check the output of ``abook --formats`` for the
file formats supported by your version of abook_.

Contributors
------------

I'd like to thank the following people who have contributed to this repository.

Ideas
'''''

* Morgan Lane

If I’ve forgotten to include your name I wholeheartedly apologise.  Just drop me
an mail_ and I’ll update the list!

Bugs
----

If you find any problems, bugs or just have a question about this package either
file an issue_ or drop me a mail_.

If you’ve found a bug please attempt to include a minimal testcase so I can
reproduce the problem, or even better a patch!

.. _abook: http://abook.sourceforge.net/
.. _GPL v3: http://www.gnu.org/licenses/
.. _Python: http://www.python.org/
.. _mail: jnrowe@gmail.com
.. _issue: https://github.com/JNRowe/blanco/issues
.. _ini: http://www.cloanto.com/specs/ini/


NEWS.rst
``blanco``
==========

User-visible changes
--------------------

.. contents::

0.6.0 - 2014-01-28
------------------

* Now defaults to ``frequency`` field in addressbook, favouring abook_ 0.6’s new
  custom field support
* arrow_ is now required
* New support files to handle dependencies for pip_ users in ``extra/``
* PEP-3101_ format support
* The testsuite now requires expecter_ and nose2_
* i18n support, open a PR with your translations now! ;)

.. _abook: http://abook.sourceforge.net/
.. _arrow: https://pypi.python.org/pypi/arrow/
.. _pip: https://pypi.python.org/pypi/pip/
.. _pep-3101: http://www.python.org/dev/peps/pep-3101/
.. _expecter: https://pypi.python.org/pypi/expecter/
.. _nose2: https://pypi.python.org/pypi/nose2/

0.5.0 - 2010-05-16
------------------

* Full documentation using Sphinx_
* ``--no-`` prefix options for overriding configuration file from the command
  line
* Basic ``setup.py`` script for installation, using distutils_

.. _Sphinx: http://sphinx.pocoo.org/
.. _distutils: http://docs.python.org/library/distutils.html

0.4.0 - 2010-05-02
------------------

* Popup notifications support, for use in desktop environments, using
  notify-python_

.. _notify-python: http://www.galago-project.org/

0.3.0 - 2010-02-19
------------------

* Support for setting command line options in ``~/.blancorc``, the format is
  ``<option>=<value>`` with no leading dashes in option names
* Much faster processing for large sent mailboxes
* Support for reading msmtp_ logs as an alternative to parsing sent mail
* Optional gmail_-specific log parser for increased accuracy in log parsing

.. _msmtp: http://msmtp.sourceforge.net/
.. _gmail: http://mail.google.com/

0.2.0 - 2010-02-11
------------------

* Support for optionally checking ``CC`` and ``BCC`` fields for last contacts
* User defined abook field for storing frequency information
* Contacts with multiple addresses are now supported
* Support for sent mail in maildir and MH formats
* Coloured output with termstyle_

.. _termstyle: https://github.com/gfxmonk/termstyle

0.1.0 - 2010-02-10
------------------

* Initial release


test/docrunner.py
#! /usr/bin/python -tt
#
"""docrunner - Execute shell tests"""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

__version__ = "0.1.0"
__date__ = "2010-02-12"
__author__ = "James Rowe <jnrowe@gmail.com>"
__copyright__ = "Copyright (C) 2010-2014  James Rowe <jnrowe@gmail.com>"
__license__ = "GNU General Public License Version 3"
__credits__ = ""
__history__ = "See git repository"

try:
    from email.utils import parseaddr
except ImportError:  # Python 2.4
    from email.Utils import parseaddr

__doc__ += """.

Run shell examples in reST literal blocks.

.. moduleauthor:: `%s <mailto:%s>`__
""" % parseaddr(__author__)

import doctest
import os
import sys

import shelldoctest as sd

if __name__ == "__main__":
    os.chdir(os.path.abspath(__file__).rsplit(os.sep, 2)[0])

    sys.exit(doctest.testfile("README.rst", module_relative=False,
                              extraglobs={"system_command": sd.system_command},
                              parser=sd.ShellDocTestParser())[0])


tests/test_blanco.py
#
"""test_blanco - Test blanco functionality"""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#


from datetime import date

from io import StringIO

from hiro import Timeline
from pytest import (mark, raises)

from blanco import (Contact, Contacts, parse_duration, parse_msmtp,
                    parse_sent, process_config, pynotify, show_note)


def test_missing_mailbox():
    with raises(IOError) as err:
        parse_sent('None')
    assert str(err.value) == "Sent mailbox ‘None’ not found"


def test_unknown_mailbox_format():
    with raises(ValueError) as err:
        parse_sent('/dev/null')
    assert str(err.value) == "Unknown mailbox format for ‘/dev/null’"


@mark.parametrize('mbox, recipients, addresses, result', [
    ('tests/data/sent.maildir', True, None, {
        'nobody@example.com': date(2000, 2, 9),
        'max@example.com': date(2000, 2, 9),
        'test@example.com': date(2010, 2, 9),
        'steven@example.com': date(2000, 2, 9),
        'joe@example.com': date(2000, 2, 9),
    }),
    ('tests/data/sent.mh', False, None, {
        'nobody@example.com': date(2000, 2, 9),
        'max@example.com': date(2000, 2, 9),
        'test@example.com': date(2010, 2, 9),
        'joe@example.com': date(2000, 2, 9),
    }),
    ('tests/data/sent.mbox', False, 'joe@example.com', {
        'joe@example.com': date(2000, 2, 9),
    }),
])
def test_parse_sent(mbox, recipients, addresses, result):
    assert parse_sent(mbox, recipients, addresses) == result


def test_missing_msmtp_log():
    with raises(IOError) as err:
        parse_msmtp('None')
    assert str(err.value) == "msmtp sent log ‘None’ not found"


@mark.parametrize('log, all_recipients, addresses, gmail, result', [
    ('tests/data/sent.msmtp', False, None, False, {
        'nobody@example.com': date(2013, 2, 9),
        'test@example.com': date(2013, 2, 9),
        'joe@example.com': date(2013, 2, 9),
    }),
    ('tests/data/sent.msmtp', True, None, False, {
        'nobody@example.com': date(2013, 2, 9),
        'max@example.com': date(2013, 2, 9),
        'test@example.com': date(2013, 2, 9),
        'steven@example.com': date(2013, 2, 9),
        'joe@example.com': date(2013, 2, 9),
    }),
    ('tests/data/sent.msmtp', False, ['nobody@example.com', ], False, {
        'nobody@example.com': date(2013, 2, 9),
    }),
    ('tests/data/sent_gmail.msmtp', False, None, True, {
        'nobody@example.com': date(2000, 2, 9),
        'test@example.com': date(2010, 2, 9),
        'joe@example.com': date(2000, 2, 9),
    }),
])
def test_parse_msmtp(log, all_recipients, addresses, gmail, result):
    with Timeline().freeze(date(2014, 6, 27)):
        assert parse_msmtp(log, all_recipients, addresses, gmail) == result


def test_parse_msmtp_invalid_gmail():
    with Timeline().freeze(date(2014, 6, 27)):
        with raises(ValueError) as err:
            parse_msmtp('tests/data/sent.msmtp', False, None, True)
        assert 'not in gmail format' in str(err.value)


def test_invalid_duration():
    with raises(ValueError) as err:
        parse_duration('1 k')
    assert str(err.value) == "Invalid duration value ‘1 k’"


@mark.parametrize('duration, result', [
    ('1d', 1),
    ('1 d', 1),
    ('0.5 y', 182),
    ('0.5 Y', 182),
])
def test_parse_duration(duration, result):
    assert parse_duration(duration) == result


def test_process_config(monkeypatch):
    monkeypatch.setattr('jnrbase.xdg_basedir.user_config',
                        lambda s: 'tests/data/valid')
    conf = process_config()
    assert conf.get('colour') is False


def test_process_config_invalid(monkeypatch):
    monkeypatch.setattr('jnrbase.xdg_basedir.user_config',
                        lambda s: 'tests/data/invalid')
    with raises(SyntaxError) as err:
        process_config()
    assert 'Invalid configuration file' in str(err.value)


class TestShowNote:
    @classmethod
    def setup_class(cls):
        cls.contact1 = Contact('James Rowe', 'jnrowe@gmail.com', 200)

    @mark.parametrize('urgency', [
        pynotify.URGENCY_NORMAL,
        pynotify.URGENCY_CRITICAL,
    ])
    def test_show_note(self, urgency, capsys):
        show_note(False, 'Note for {}', self.contact1, urgency=urgency)
        out, _ = capsys.readouterr()
        # Use contains to avoid having to mess around with {,no-}colour options
        assert 'Note for James Rowe' in out

    @mark.parametrize('show_succeeds', [True, False])
    def test_show_note_pynotify(self, show_succeeds, monkeypatch):
        class Notification:
            titles = bodies = icons = []

            def __init__(self, t, s, i):
                self.titles.append(t)
                self.bodies.append(s)
                self.icons.append(i)

            def set_urgency(self, u):
                self.u = u

            def set_timeout(self, o):
                self.o = o

            def show(self):
                return True
        monkeypatch.setattr(pynotify, 'Notification', Notification,
                            raising=False)
        monkeypatch.setattr(pynotify, 'get_server_caps',
                            staticmethod(lambda: []), raising=False)
        if show_succeeds:
            show_note(True, 'Note for {}', self.contact1)
            assert 'Note for James Rowe' in Notification.bodies
        else:
            monkeypatch.setattr(pynotify.Notification, 'show', lambda s: False)
            with raises(OSError) as err:
                show_note(True, 'Broken note for {}', self.contact1)
            assert str(err.value) == 'Notification failed to display!'


class TestContact:
    @classmethod
    def setup_class(cls):
        cls.contact1 = Contact('James Rowe', 'jnrowe@gmail.com', 200)
        cls.contact2 = Contact('James Rowe',
                               ['jnrowe@gmail.com', 'jnrowe@example.com'],
                               200)

    def test___repr__(self):
        assert repr(self.contact1) == \
            "Contact('James Rowe', ['jnrowe@gmail.com'], 200)"

    @mark.parametrize('contact, expected', [
        ('contact1', 'James Rowe <jnrowe@gmail.com> (200 days)'),
        ('contact2', 'James Rowe <jnrowe@gmail.com, jnrowe@example.com> '
                     '(200 days)'),
    ])
    def test___str__(self, contact, expected):
        assert str(getattr(self, contact)) == expected

    @mark.parametrize('contact, spec, expected', [
        ('contact1', '', 'James Rowe <jnrowe@gmail.com> (200 days)'),
        ('contact1', 'email', 'James Rowe <jnrowe@gmail.com>'),
        ('contact2', 'email', 'James Rowe <jnrowe@gmail.com>'),
    ])
    def test___format__(self, contact, spec, expected):
        assert format(getattr(self, contact), spec) == expected

    def test___format___invalid_type(self):
        with raises(ValueError) as err:
            format(self.contact1, 'hat style')
        assert str(err.value) == "Unknown format_spec 'hat style'"

    def test_trigger(self):
        assert self.contact1.trigger({
            'jnrowe@gmail.com': date(1942, 1, 1),
        }) == date(1942, 7, 20)

    @mark.parametrize('server_caps, expected', [
        ([], 'James Rowe'),
        (['body-hyperlinks'],
         "<a href='mailto:jnrowe@gmail.com'>James Rowe</a>"),
    ])
    def test_notify_str(self, server_caps, expected, monkeypatch):
        monkeypatch.setattr(pynotify, 'get_server_caps',
                            staticmethod(lambda: server_caps), raising=False)
        assert self.contact1.notify_str() == expected


class TestContacts:
    def test___repr__(self):
        assert repr(Contacts([
            Contact('James Rowe', 'jnrowe@gmail.com', 200),
        ])) == \
            "Contacts([Contact('James Rowe', ['jnrowe@gmail.com'], 200)])"

    def test_addresses(self):
        p = Contacts([
            Contact('Bill', ['test@example.com', 'new@example.com'], 30),
            Contact('Joe', ['joe@example.com'], 30)
        ])
        assert p.addresses() == ['test@example.com', 'new@example.com',
                                 'joe@example.com']

    def test_parse(self):
        contacts = Contacts()
        contacts.parse('tests/data/blanco.conf', 'frequency')

        assert repr(contacts) == \
            ('Contacts(['
             "Contact('Bill', ['test@example.com'], 30), "
             "Contact('Joe', ['joe@example.com'], 30), "
             "Contact('Steven', ['steven@example.com'], 365)"
             '])')

    def test_parse_missing_file(self, tmpdir):
        contacts = Contacts()
        with raises(IOError) as err:
            contacts.parse(str(tmpdir.join('no_such_file')), 'frequency')
        assert 'Addressbook file not found' in str(err.value)


blanco.py
#! /usr/bin/env python
"""blanco - Keep in touch, barely."""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import sys

from blanco import main


sys.exit(main())


blanco/i18n.py
#
# coding=utf-8
"""i18n - Internationalisation support for blanco"""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# pylint: disable-msg=C0121

import gettext

from os import path

PACKAGE_LOCALE = path.join(path.realpath(path.dirname(__file__)), 'locale')

gettext.install('blanco', PACKAGE_LOCALE)

_, N_ = gettext.gettext, gettext.ngettext


blanco/compat.py
#
# coding=utf-8
"""compat - Python 2/3 compatibility support for blanco"""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from sys import version_info

PY2 = version_info[0] == 2

if PY2:
    import codecs

    basestring = basestring
    open = codecs.open
else:
    basestring = str
    open = open

if PY2:
    def mangle_repr_type(klass):
        klass.__repr_unicode__ = klass.__repr__

        def wrapper(self):
            return self.__repr_unicode__().encode('utf-8')
        klass.__repr__ = wrapper
        return klass
else:
    mangle_repr_type = lambda x: x


extra/docrunner.py
#! /usr/bin/env python
"""docrunner - Execute shell tests"""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

__doc__ += """.

Run shell examples in reST literal blocks."""

import doctest
import sys

import shelldoctest as sd

if __name__ == '__main__':
    sys.exit(doctest.testfile(sys.argv[1], module_relative=False,
                              extraglobs={'system_command': sd.system_command},
                              parser=sd.ShellDocTestParser())[0])


blanco/__init__.py
#! /usr/bin/python -tt
"""blanco - Keep in touch, barely.

Check sent mail to make sure you’re keeping in contact with your friends.
"""
# Copyright © 2010-2014  James Rowe <jnrowe@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from . import _version

__version__ = _version.dotted
__date__ = _version.date
__author__ = 'James Rowe <jnrowe@gmail.com>'
__copyright__ = 'Copyright (C) 2010-2014  James Rowe <jnrowe@gmail.com>'
__license__ = 'GNU General Public License Version 3'
__credits__ = ''
__history__ = 'See git repository'

import datetime
import errno
import mailbox
import operator
import os
import re
import sys
import time

from email.utils import (formataddr, getaddresses, parsedate)

import click
import configobj
import validate

try:
    import pynotify
except ImportError:
    class _Fake_PyNotify:  # NOQA
        URGENCY_CRITICAL = 2
        URGENCY_NORMAL = 1
        URGENCY_LOW = 0
        EXPIRES_DEFAULT = -1
        EXPIRES_NEVER = 0
    pynotify = _Fake_PyNotify  # NOQA

from jnrbase import (colourise, i18n, xdg_basedir)


_, N_ = i18n.setup(_version)


USAGE = _("Check sent mail to make sure you’re keeping in contact with your "
          "friends.")


def parse_sent(path, all_recipients=False, addresses=None):
    """Parse sent messages mailbox for contact details.

    :param str path: Location of the sent mailbox
    :param bool all_recipients: Whether to include CC and BCC addresses in
        results, or just the first
    :param list addresses: Addresses to look for in sent mail, all if not
        specified
    :rtype: `dict` of `str` keys and `datetime.datetime` values
    :return: Keys of email address, and values of seen date
    """
    path = os.path.expanduser(path)
    if not os.path.exists(path):
        raise IOError(_('Sent mailbox ‘{}’ not found').format(path))
    if os.path.isdir('{}/new'.format(path)):
        mtype = mailbox.Maildir
    elif os.path.exists('{}/.mh_sequences'.format(path)):
        mtype = mailbox.MH
    elif os.path.isfile(path):
        mtype = mailbox.mbox
    else:
        raise ValueError(_('Unknown mailbox format for ‘{}’').format(path))
    # Use factory=None to work around the rfc822.Message default for Maildir.
    mbox = mtype(path, factory=None, create=False)

    contacts = []
    for message in mbox:
        fields = message.get_all('to', [])
        if all_recipients:
            fields.extend(message.get_all('cc', []))
            fields.extend(message.get_all('bcc', []))
        results = [x[1].lower() for x in getaddresses(fields)]
        date = datetime.datetime(*parsedate(message['date'])[:-2])
        contacts.extend([(address, date.date()) for address in results
                         if not addresses or address in addresses])
    return dict(sorted(contacts, key=operator.itemgetter(1)))


def parse_msmtp(log, all_recipients=False, addresses=None, gmail=False):
    """Parse sent messages mailbox for contact details.

    :param str log: Location of the msmtp logfile
    :param bool all_recipients: Whether to include all recipients in results,
        or just the first
    :param list addresses: Addresses to look for in sent mail, all if not
        specified
    :param bool gmail: Log is for a gmail account
    :rtype: `dict` of `str` keys and `datetime.datetime` values
    :return: Keys of email address, and values of seen date
    """
    if not os.path.exists(log):
        raise IOError(_('msmtp sent log ‘{}’ not found').format(log))

    matcher = re.compile('recipients=([^ ]+)')
    gmail_date = re.compile('smtpmsg.*OK ([^ ]+)')

    start = datetime.datetime.utcfromtimestamp(os.path.getmtime(log))

    year = start.year
    md = start.month, start.day
    contacts = []
    for line in reversed([line for line in open(log)
                          if line.endswith('exitcode=EX_OK\n')]):
        if gmail:
            gd = gmail_date.search(line)
            try:
                ts = int(gd.groups()[0])
                parsed = datetime.datetime.utcfromtimestamp(ts)
            except AttributeError:
                raise ValueError(
                    _('msmtp {!r} log is not in gmail format').format(log))
            year = parsed.year
            md = parsed.month, parsed.day
        else:
            date = time.strptime(line[:6], '%b %d')[1:3]
            if date > md:
                year = year - 1
            md = date

        recips = matcher.search(line, re.DOTALL)
        results = [s.lower() for s in recips.groups()[0].split(',')]
        if not all_recipients:
            results = [results[0], ]
        contacts.extend([(address, datetime.datetime(year, *md).date())
                         for address in results
                         if not addresses or address in addresses])
    # Sorting prior to making the dictionary means we only use the latest
    # entry.
    return dict(sorted(contacts, key=operator.itemgetter(1)))


def parse_duration(duration):
    """Parse human readable duration.

    :param str duration: Duration definition
    :rtype: `int`
    :return: Number of days in ``duration``
    :raise ValueError: Invalid value for ``duration``
    """
    match = re.match('^(\d+(?:|\.\d+)) *([dwmy])$', duration, re.IGNORECASE)
    if not match:
        raise ValueError(_('Invalid duration value ‘{}’').format(duration))
    value, units = match.groups()
    units = 'dwmy'.index(units.lower())
    # days per day/week/month/year
    multiplier = (1, 7, 28, 365)
    return int(float(value) * multiplier[units])


def process_config():
    """Main configuration file.

    :rtype: `configobj.ConfigObj`
    :return: Parsed configuration file
    """
    config_file = os.path.join(xdg_basedir.user_config('blanco'), 'config.ini')
    config_spec = [
        'colour = boolean(default=True)',
        "addressbook = string(default='~/.abook/addressbook')",
        "sent type = string(default='mailbox')",
        'all = boolean(default=False)',
        "mbox = string(default='~/Mail/Sent')",
        "log = string(default='~/Mail/.logs/gmail.log')",
        'gmail = boolean(default=True)',
        "field = string(default='frequency')",
        'notify = boolean(default=False)',
        'verbose = boolean(default=False)',
    ]
    config = configobj.ConfigObj(config_file, configspec=config_spec)
    results = config.validate(validate.Validator())
    if results is not True:
        for key in results:
            if results[key]:
                continue
            colourise.pfail(_('Config value for {!r} is invalid').format(key))
        raise SyntaxError(
            _('Invalid configuration file {!r}').format(config_file))

    return config


def show_note(notify, message, contact, urgency=pynotify.URGENCY_NORMAL,
              expires=pynotify.EXPIRES_DEFAULT):
    """Display reminder.

    :param bool notify: Whether to use notification popup
    :param str message: Message string to show
    :param Contact contact: Contact to show message for
    :param int urgency: Urgency state for message
    :param int expires: Time to show notification popup in milliseconds
    :raise OSError: Failure to show notification
    """
    if notify:
        note = pynotify.Notification(_('Hey, remember me?'),
                                     message.format(contact.notify_str()),
                                     'stock_person')
        note.set_urgency(urgency)
        note.set_timeout(expires)

        if not note.show():
            raise OSError(_('Notification failed to display!'))
    else:
        if urgency == pynotify.URGENCY_CRITICAL:
            colourise.pfail(message.format(contact.name))
        else:
            colourise.pwarn(message.format(contact.name))


class Contact:

    """Simple contact class."""

    def __init__(self, name, addresses, frequency):
        """Initialise a new `Contact` object."""
        self.name = name
        if isinstance(addresses, str):
            self.addresses = [addresses.lower(), ]
        else:
            self.addresses = [s.lower() for s in addresses]
        self.frequency = frequency

    def __repr__(self):
        """Self-documenting string representation."""
        return '{}({!r}, {!r}, {!r})'.format(self.__class__.__name__,
                                             self.name, self.addresses,
                                             self.frequency)

    def __str__(self):
        """Pretty printed contact string."""
        return '{} <{}> ({} days)'.format(self.name, ', '.join(self.addresses),
                                          self.frequency)

    def __format__(self, format_spec):
        """Extended pretty printing for `Contact` strings.

        :param str format_spec: Coordinate formatting system to use
        :rtype: `str`
        :return: Human readable string representation of `Contact` object
        :raise ValueError: Unknown value for ``format_spec``
        """
        if not format_spec:  # default format calls set format_spec to ''
            return str(self)
        elif format_spec == 'email':
            return formataddr((self.name, self.addresses[0]))
        else:
            raise ValueError(_('Unknown format_spec {!r}').format(format_spec))

    def trigger(self, sent):
        """Calculate trigger date for contact.

        :type sent: `dict` of `str` keys and `datetime.datetime values
        :param sent: Address to last seen dictionary
        :rtype: `datetime.datetime`
        :return: Date to start reminders on
        """
        match = sorted([v for k, v in sent.items() if k in self.addresses])[0]
        return match + datetime.timedelta(days=self.frequency)

    def notify_str(self):
        """Calculate trigger date for contact.

        :rtype: `str`
        :return: Stylised name for use with notifications
        """
        if 'body-hyperlinks' in pynotify.get_server_caps():
            name = "<a href='mailto:{}'>{}</a>".format(self.addresses[0],
                                                       self.name)
        else:
            name = self.name
        return name


class Contacts(list):

    """Group of `Contact`."""

    def __init__(self, contacts=None):
        """Initialise a new `Contacts` object."""
        super(Contacts, self).__init__()
        if contacts:
            self.extend(contacts)

    def __repr__(self):
        """Self-documenting string representation."""
        return '{}({!r})'.format(self.__class__.__name__,
                                 sorted(self[:],
                                        key=operator.attrgetter('name')))

    def addresses(self):
        """Fetch all addresses of all `Contact` objects.

        :rtype: `list` of `str`
        :return: Addresses of every `Contact`
        """
        return [address for contact in self for address in contact.addresses]

    def parse(self, addressbook, field):
        """Parse address book for usable entries.

        :param str addressbook: Location of the address book to useful
        :param str field: Address book field to use for contact frequency
        """
        if not os.path.isfile(addressbook):
            raise IOError('Addressbook file not found {!r}'.format(addressbook))
        config = configobj.ConfigObj(os.path.expanduser(addressbook))

        for entry in config.values():
            if not field in entry:
                continue
            self.append(Contact(entry['name'], entry['email'],
                                parse_duration(entry[field])))


@click.command(help=USAGE, epilog=_('Please report bugs to jnrowe@gmail.com'))
@click.option('-a', '--addressbook', metavar='FILENAME',
              help=_('Address book to read contacts from.'))
@click.option('-t', '--sent-type', type=click.Choice(['mailbox', 'msmtp']),
              help=_('Sent source type.'))
@click.option('-r', '--all/--no-all',
              help=_('Include all recipients(CC and BCC fields).'))
@click.option('-m', '--mbox', metavar='FILENAME',
              help=_('Mailbox used to store sent mail.'))
@click.option('-l', '--log', metavar='FILENAME', help=_('msmtp log to parse.'))
@click.option('-g', '--gmail/--no-gmail',
              help=_('Log from a gmail account(use accurate filter).'))
@click.option('-s', '--field',
              help=_('Addressbook field to use for frequency value.'))
@click.option('-n', '--notify/--no-notify',
              help=_('Display reminders using notification popups.'))
@click.option('--colour/--no-colour', envvar='BLANCO_COLOUR', default=None,
              help=_('Output colourised informational text.'))
@click.option('-v', '--verbose/--no-verbose',
              help=_('Produce verbose output.'))
@click.version_option(_version.dotted)
def main(addressbook, sent_type, all, mbox, log, gmail, field, notify, colour,
         verbose):  # pragma: no cover
    """Main script."""
    config = process_config()
    if not addressbook:
        addressbook = os.path.expanduser(config.get('addressbook'))
    if not sent_type:
        sent_type = config.get('sent type')
    if not all:
        all = config.as_bool('all')
    if not mbox:
        mbox = os.path.expanduser(config.get('mbox'))
    if not log:
        log = os.path.expanduser(config.get('log'))
    if not gmail:
        gmail = config.as_bool('gmail')
    if not field:
        field = config.get('field')
    if not notify:
        notify = config.as_bool('notify')
    if not verbose:
        verbose = config.as_bool('verbose')
    if colour is None:
        if 'color' in config:
            config['colour'] = config['color']
        colour = config.as_bool('colour')
    colourise.COLOUR = colour

    if notify and pynotify is _Fake_PyNotify:
        raise click.UsageError(
            colourise.fail(_('Notification popups require the notify-python '
                             'package') + '\n'))

    if notify:
        if not pynotify.init(sys.argv[0]):
            colourise.pfail(_('Unable to initialise pynotify!'))
            return errno.EIO

    contacts = Contacts()
    contacts.parse(addressbook, field)
    try:
        if sent_type == 'msmtp':
            sent = parse_msmtp(log, all, contacts.addresses(), gmail)
        else:
            sent = parse_sent(mbox, all, contacts.addresses())
    except IOError as e:
        colourise.pfail(e.args[0])
        return errno.EPERM

    now = datetime.datetime.utcnow()
    for contact in contacts:
        if not any(address in sent for address in contact.addresses):
            show_note(notify, _('No mail record for {}'), contact)
        elif now > contact.trigger(sent):
            show_note(notify, _('Mail due for {}'), contact,
                      pynotify.URGENCY_CRITICAL, pynotify.EXPIRES_NEVER)
